### PR TITLE
fix: allow custom framework definition in cypress.config.ts

### LINKED ISSUE #26394: Using a community/custom component framework definition causes a TypeScript error
### Current behavior

When you use an unofficial framework in your cypress.config.ts, TypeScript complains.

### Desired behavior

You should be allowed to use framework definitions that fit the format `cypress-ct-[name]` or `@[scope]/cypress-ct-[name]`.

### Test code to reproduce

cypress.config.ts
```ts
import { defineConfig } from 'cypress';
import { defineConfig as defineViteConfig } from 'vite';

export default defineConfig({
  component: {
    devServer: {
      framework: 'cypress-ct-lit-element', // Error on this line
      bundler: 'vite'
    }
  }
});

```

### Cypress Version

12.9.0

### Node version

16.16.0

### Operating System

MacOS 13.2.1

### Debug Logs

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by kgroat**
- Closes https://github.com/cypress-io/cypress/issues/26394

### Additional details

According to [the docs](https://docs.cypress.io/guides/component-testing/third-party-definitions#Framework-Definition) you can use a custom framework definition that matches either `cypress-ct-*` or `@organization/cypress-ct-*` but the TypeScript types don't support this yet.  This PR allows those custom definitions.

### Steps to test

Create a cypress.config.ts with the following content:
```ts
import { defineConfig } from 'cypress';
import { defineConfig as defineViteConfig } from 'vite';

export default defineConfig({
  component: {
    devServer: {
      framework: 'cypress-ct-lit-element', // Error on this line
      bundler: 'vite'
    }
  }
});
```

### How has the user experience changed?

The user can now use custom framework definitions in TypeScript config files.

### PR Tasks

- [na] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by lmiller1990** [comment]
Looks good to me, thanks! I pushed https://github.com/cypress-io/cypress/pull/26396/commits/dfe125b44a8f58f5903ae0f0436d5d880d9dc924 to add some tests.  💯

**Review by lmiller1990 — APPROVED** [review]
Looks good, I merged in develop and added some tests for the type definitions. 💯 

Hmm linting is not happy, I think because `${}` in strings is a new-ish TS feature, we only require 3.4+. Anyway, I think we can still do this, just need to figure out the linting situation...  users on older versions won't get the same level of type safety, `cypress-ct-*` will error for them. I don't think we can do much about that, though.

**Comment by lmiller1990** [comment]
`lint-types` is complaining - I cannot seem to run this step locally, or figure out what it wants.

**Comment by astone123** [comment]
Yeah for some reason the use of the template literal syntax [here](https://github.com/cypress-io/cypress/pull/26396/files#diff-9fde1fb2f7eaba5ea49b585c811461495bfbba88e614e900a96863ecfe6b980aR3471) is causing `dtslint` to throw a bunch of other unrelated type errors for TS 4.0. Adding the "minimum version" comment doesn't appear to fix this.

**Comment by kgroat** [comment]
Yeah, it's very odd that it's causing issues in completely separate files even.  I'm not sure why that would be happening.

I tried approaching this from a couple of different ways now, and all of them give the same errors when running the `dtslint` script.  I'm not sure if there is a way forward.

**Review by jordanpowell88 — APPROVED** [review]
Thank you!

**Comment by lmiller1990** [comment]
dtslint is really painful to debug. It won't even run on machine.

We will pull this into our next sprint of work (start next week) so someone can go deep and figure it out, if no-one can figure it out before then.

**Comment by kgroat** [comment]
Not sure if it would make a difference, but the `dtslint` package appears to have been moved to `@definitelytyped/dtslint`:
https://www.npmjs.com/package/@definitelytyped/dtslint

Maybe upgrading would solve the issue?  Not sure what the upgrade path looks like though.

**Comment by lmiller1990** [comment]
Let's find out! I don't think I can get anyone on this internally until next week, though. If you'd like to take a stab at updating dts, that'd be great (or finding a solution so the lint step stops failing - not sure what that looks like, though...)

File: cli/CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +17,7 @@ _Released 04/11/2023 (PENDING)_
 
 - Upgraded [`minimist`](https://www.npmjs.com/package/minimist) from `1.2.6` to `1.2.8` to address this [CVE-2021-44906](https://github.com/advisories/GHSA-xvch-5gv4-984h) NVD security vulnerability. Addressed in [#26254](https://github.com/cypress-io/cypress/pull/26254).
 
+
```
Comment (emilyrohrbough) [inline cli/CHANGELOG.md:36]:
```suggestion
```

File: cli/CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,7 @@ _Released 04/11/2023 (PENDING)_
 
  - Capture the [Azure](https://azure.microsoft.com/) CI provider's environment variable [`SYSTEM_PULLREQUEST_PULLREQUESTNUMBER`](https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#system-variables-devops-services) to display the linked PR number in the Cloud. Addressed in [#26215](https://github.com/cypress-io/cypress/pull/26215).
  - Fixed an issue in the onboarding wizard where project framework & bundler would not be auto-detected when opening directly into component testing mode using the `--component` CLI flag. Fixes [#22777](https://github.com/cypress-io/cypress/issues/22777).
+ - Fixed an issue where the TypeScript types did not support [custom framework definitions](https://docs.cypress.io/guides/component-testing/third-party-definitions#Framework-Definition) in cypress.config.ts.  Fixes [#26394](https://github.com/cypress-io/cypress/issues/26394).
```
Comment (emilyrohrbough) [inline cli/CHANGELOG.md]:
```suggestion
 - Fixed an issue where the TypeScript types did not support [custom framework definitions](https://docs.cypress.io/guides/component-testing/third-party-definitions#Framework-Definition) in cypress.config.ts. Fixes [#26394](https://github.com/cypress-io/cypress/issues/26394).
```

File: cli/types/cypress.d.ts [hunk 3]
```diff
@@ -2,6 +2,7 @@
 /// <reference path="./cypress-npm-api.d.ts" />
 /// <reference path="./cypress-eventemitter.d.ts" />
 /// <reference path="./cypress-type-helpers.d.ts" />
+// Minimum TypeScript Version: 4.1
```
Comment (emilyrohrbough) [inline cli/types/cypress.d.ts]: If this places a minimum typescript requirement in our shipped types, what happens if someone is on an older version? Will this cause issues? Wondering how passive this is.

Comment (lmiller1990) [reply cli/types/cypress.d.ts]: Trying to figure this out still... I wonder if we can bump our minimum TS version? 5.x is out now -- 3.4 came out in Mar 2019.

Comment (kgroat) [reply cli/types/cypress.d.ts]: FWIW, the [Definitely Typed support window](https://github.com/DefinitelyTyped/DefinitelyTyped#support-window) is 4.3 - 5.0 (they support each version for 2 years from its release).

This is also the set of versions checked by the latest [`@definitelytyped/dtslint`](https://www.npmjs.com/package/@definitelytyped/dtslint)

Comment (lmiller1990) [reply cli/types/cypress.d.ts]: Oh ok, that might be the issue.

I think we might want to look into bumping out minimal TS version as well. We will pick this up in our next sprint of work (2 week blocks, starting new one tomorrow). So this won't be in the next release (tomorrow) but the following one (2 weeks time).

Comment (emilyrohrbough) [reply cli/types/cypress.d.ts]: @lmiller1990 FYI this was logged to bump our minimum version & flagged as breaking: https://github.com/cypress-io/cypress/issues/24875

Comment (lmiller1990) [reply cli/types/cypress.d.ts]: I wonder if we can go to 4.x, considering 5.x is now out.

I'm not entirely clear on what makes this a breaking change - any context?

Comment (lmiller1990) [reply cli/types/cypress.d.ts]: Posted in the issue: https://github.com/cypress-io/cypress/issues/24875#issuecomment-1507662491

**Comment by astone123** [comment]
@kgroat I just messed this branch up by accident - rebase didn't go as expected. I'm not sure how to get it back to how it was before. Mind if I create a new branch on your fork and make a new PR?

**Comment by astone123** [comment]
Closing in favor of https://github.com/cypress-io/cypress/pull/26545

**Pull request closed without merging**